config/module.go

```diff
@@ -213,3 +214,80 @@ func (m Module) EvaluateExePath(packagesDir string) (string, error) {
 	}
 	return m.ExePath, nil
 }
+
+// FirstRunSuccessExt is the extension of the file whose existence
+// denotes that the setup phase for a module ran successfully.
+const FirstRunSuccessExt = ".first_run_succeeded"
+
+// EvaluateFirstRunPath returns absolute FirstRunPath from one of two sources (in order of precedence):
+// 1. if there is a meta.json in the exe dir, use that, except in local non-tarball case.
+// 2. if this is a local tarball and there's a meta.json next to the tarball, use that.
+// Note: the working directory must be the unpacked tarball directory or local exec directory.
+//
+// On success (i.e. if the returned error is nil), this function also returns a function that creates
+// a marker file indicating that the setup phase has run successfully.
+func (m Module) EvaluateFirstRunPath(packagesDir string) (
+	string,
+	func() error,
+	error,
+) {
+	noop := func() error { return nil }
+	firstRunDir, err := m.exeDir(packagesDir)
+	if err != nil {
+		return "", noop, err
+	}
+
+	firstRunSuccessPath := firstRunDir + FirstRunSuccessExt
+	if _, err := os.Stat(firstRunSuccessPath); !errors.Is(err, os.ErrNotExist) {
```

Comment:
nit: organizationally, I don't think the evaluation of first run success should be in a `EvaluateFirstRunPath` function - it feels like the function you would run before this function, since you wouldn't need to find the first run path any more at this point. However, I do see a lot of vars are shared so I can see this being fine as well

Reply:
> I do see a lot of vars are shared so I can see this being fine as well

yep, this is why i organized it the way you see it. your suggestion is how i would want it to be organized also. 

i will merge this as is and try to re-arrange when addressing all of the fast-follows for this feature